logical/plugin/plugin.go

```diff
@@ -124,3 +132,23 @@ func newPluginClient(sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginR
 		backendPluginClient: backendRPC,
 	}, nil
 }
+
+// wrapError takes a generic error type and makes it usable with the plugin
+// interface. Only errors which have exported fields and have been registered
+// with gob can be unwrapped and transported. This checks error types and, if
+// none match, wrap the error in a plugin.BasicError.
+func wrapError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	switch err.(type) {
+	case *plugin.BasicError,
+		logical.HTTPCodedError,
+		*logical.StatusBadRequest,
+		*logical.ReplicationCodedError:
```

Comment:
Maybe we should add `logical.ErrUnsupportedOperation` here (and above) as well. That will allow us to remove this block: 
https://github.com/hashicorp/vault/blob/master/logical/plugin/backend_client.go#L117-L119

Reply:
So, from what I understand, you can't gob a thing without public fields, and I don't think `errors.New` [creates a struct with any all private fields](https://golang.org/src/errors/errors.go?s=353:520#L1). I believe Gob will panic if you give it that error.

Reply:
Yea - I just tested to make sure: 

```
2017/10/19 13:33:15.802958 [DEBUG] plugin.vault-auth-slack: 2017/10/19 13:33:15 rpc: gob error encoding body: gob: type errors.errorString has no exported fields
```

So we can't actually gob that error unless we choose to convert it to a concrete type with exported fields like:

```go
type ErrUnsupportedOperation struct {}

func (e *ErrUnsupportedOperation) Error() string {
  return "Unsupported operation"
}
```

Then it would be gobbable, _but_ that's probably a breaking change, since you can no longer compare with `==` (have to use a type switch):

```go
switch err.(type) {
case *ErrUnsupportedOperation:
default:
  # ...
}
```

Reply:
Can't we do it if we have something like:

```go
type UnsupportedOperationError struct {}

func (e *UnsupportedOperationError) Error() string {
  return "Unsupported operation"
}

var ErrUnsupportedOperation = &UnsupportedOperationError{}
```

We would then register  `UnsupportedOperationError`. When we are using error `err` returned back from HandleRequest on checks such as `errwrap.Contains(err, ErrUnsupportedOperation.Error())` should still work right? However I am fine to keep it the way it is currently, too.

Reply:
No - you need (public) exported fields. Gob is weird